posix: implement pthread_attr_getinheritsched() and pthread_attr_setinheritsched()

moonlight83340 opened the pull request:
This is part of the See https://github.com/zephyrproject-rtos/zephyr/issues/51211 (RFC https://github.com/zephyrproject-rtos/zephyr/issues/51211).

`pthread_attr_setinheritsched()` and `pthread_attr_getinheritsched()` are required as part of _POSIX_THREAD_PRIORITY_SCHEDULING Option Group.

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_attr_setinheritsched.html

Fixes  #66968
Fixes #66966

moonlight83340 committed:
doc: posix: mark set and get inheritsched as supported

`pthread_attr_getinheritsched()` and
`pthread_attr_setinheritsched()` are now implemented,
mark it so.

signed-off-by: Gaetan Perrot <gaetanperrotpro@gmail.com>

moonlight83340 committed:
posix: Implement tests for set and get inheritsched APIs for pthread attr

Add tests for `pthread_attr_setinheritsched()`
and `pthread_attr_getinheritsched()`

signed-off-by: Gaetan Perrot <gaetanperrotpro@gmail.com>

moonlight83340 commented:
What should I set on initial value in `pthread_attr_init()` ?
https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_attr_setinheritsched.html
I may haven't take good look at it but I see no mention 😓

ycsin commented:
> What should I set on initial value in `pthread_attr_init()` ?

It seems to be "_implementation-defined_" according to the [spec](https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_09_04):
> When a process is created, its single thread has a scheduling policy and associated attributes equal to the policy and attributes of the process. The default scheduling contention scope value is implementation-defined. The default values of other scheduling attributes are implementation-defined.

I did a quick search on google for `"default value of inheritsched"` and seems like most OSes have the default value of `PTHREAD_INHERIT_SCHED`:
- [QNX](https://www.qnx.com/developers/docs/6.5.0SP1.update/com.qnx.doc.neutrino_lib_ref/p/pthread_attr_setinheritsched.html) ("_The default value of the thread inherit scheduling attribute is PTHREAD_INHERIT_SCHED._")
- [IBM](https://www.ibm.com/docs/en/aix/7.2?topic=programming-scheduling-threads) ("_The default value of the inheritsched attribute is PTHREAD_INHERIT_SCHED_")
- [Oracle](https://docs.oracle.com/cd/E19253-01/816-5137/attrib-27410/index.html) ("_might change from PTHREAD_EXPLICIT_SCHED to PTHREAD_INHERIT_SCHED in a future Solaris release_")

moonlight83340 commented:
> > What should I set on initial value in `pthread_attr_init()` ?
> 
> It seems to be "_implementation-defined_" according to the [spec](https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_09_04):
> 
> > When a process is created, its single thread has a scheduling policy and associated attributes equal to the policy and attributes of the process. The default scheduling contention scope value is implementation-defined. The default values of other scheduling attributes are implementation-defined.
> 
> I did a quick search on google for `"default value of inheritsched"` and seems like (based on my observation in the search result) most Linux OSes have the default value of `PTHREAD_INHERIT_SCHED`:
> 
>     * [QNX Neutrino](https://www.qnx.com/developers/docs/6.5.0SP1.update/com.qnx.doc.neutrino_lib_ref/p/pthread_attr_setinheritsched.html) ("_The default value of the thread inherit scheduling attribute is PTHREAD_INHERIT_SCHED._")
> 
>     * [IBM AIX](https://www.ibm.com/docs/en/aix/7.2?topic=programming-scheduling-threads) ("_The default value of the inheritsched attribute is PTHREAD_INHERIT_SCHED_")
> 
>     * [Oracle Solaris](https://docs.oracle.com/cd/E19253-01/816-5137/attrib-27410/index.html) ("_might change from PTHREAD_EXPLICIT_SCHED to PTHREAD_INHERIT_SCHED in a future Solaris release_")
> 
> 
> The [Apache NuttX](https://github.com/apache/nuttx/blob/master/include/nuttx/pthread.h#L51C5-L51C27) however, has the default value of `PTHREAD_EXPLICIT_SCHED`

Should we go on the mainly used way ?

> > What should I set on initial value in `pthread_attr_init()` ?
> 
> It seems to be "_implementation-defined_" according to the [spec](https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_09_04):
> 
> > When a process is created, its single thread has a scheduling policy and associated attributes equal to the policy and attributes of the process. The default scheduling contention scope value is implementation-defined. The default values of other scheduling attributes are implementation-defined.
> 
> I did a quick search on google for `"default value of inheritsched"` and seems like (based on my observation in the search result) most Linux OSes have the default value of `PTHREAD_INHERIT_SCHED`:
> 
>     * [QNX Neutrino](https://www.qnx.com/developers/docs/6.5.0SP1.update/com.qnx.doc.neutrino_lib_ref/p/pthread_attr_setinheritsched.html) ("_The default value of the thread inherit scheduling attribute is PTHREAD_INHERIT_SCHED._")
> 
>     * [IBM AIX](https://www.ibm.com/docs/en/aix/7.2?topic=programming-scheduling-threads) ("_The default value of the inheritsched attribute is PTHREAD_INHERIT_SCHED_")
> 
>     * [Oracle Solaris](https://docs.oracle.com/cd/E19253-01/816-5137/attrib-27410/index.html) ("_might change from PTHREAD_EXPLICIT_SCHED to PTHREAD_INHERIT_SCHED in a future Solaris release_")
> 
> 
> The [Apache NuttX](https://github.com/apache/nuttx/blob/master/include/nuttx/pthread.h#L51C5-L51C27) however, has the default value of `PTHREAD_EXPLICIT_SCHED`

moonlight83340 commented:
closed by error 😓

cfriedt commented:
Wow @moonlight83340 - you have been very busy!

Thanks - I'm taking a look now.

cfriedt reviewed (commented):
@moonlight83340 - nice work 👍

I think there may be some additional tests to write, and likely some additional work on the implementation of `pthread_create()`

Tests:
1. Verify that the scheduling attributes of the parent thread (policy, priority, ..) are inherited by a child thread when INHERIT_SCHED is used. This may be tricky, because currently Zephyr threads are not pthreads. It might be necessary to extrapolate equivalent pthread attributes (policy, priority) from the running `k_thread` and compare those.

2. Verify that the scheduling attributes specified by the pthread_attr_t are passed to the child thread.

For both of those, you should be able to create a static helper function in the same file. Something like this?

```cpp
static void test_pthread_attr_set_inheritsched_common(int inheritsched);
```

moonlight83340 commented:
> @moonlight83340 - nice work 👍
> 
> I think there may be some additional tests to write, and likely some additional work on the implementation of `pthread_create()`
> 
> Tests:
> 
>     1. Verify that the scheduling attributes of the parent thread (policy, priority, ..) are inherited by a child thread when INHERIT_SCHED is used. This may be tricky, because currently Zephyr threads are not pthreads. It might be necessary to extrapolate equivalent pthread attributes (policy, priority) from the running `k_thread` and compare those.
> 
>     2. Verify that the scheduling attributes specified by the pthread_attr_t are passed to the child thread.
> 
> 
> For both of those, you should be able to create a static helper function in the same file. Something like this?
> 
> ```c++
> static void test_pthread_attr_set_inheritsched_common(int inheritsched);
> ```

Still working on the test part not sure on how I will do that but I think I got it for the `pthread_create() ` part.

I have done something like that :

```
	struct posix_thread *self = NULL;

	K_SPINLOCK(&pthread_pool_lock) {
		self = to_posix_thread(pthread_self());
		if (self == NULL) {
			K_SPINLOCK_BREAK;
		}
		if( (self_attr.inheritsched == PTHREAD_INHERIT_SCHED) {
			t->attr.priority = self->attr.priority;
			t->attr.schedpolicy = self->attr.schedpolicy;
			t->attr.inheritsched = self->attr.inheritsched;
		}
	}
```

moonlight83340 committed:
posix: Implement set and get inheritsched APIs for pthread attr

Implement `pthread_attr_setinheritsched()` and
`pthread_attr_getinheritsched()`are required
as part of _POSIX_THREAD_PRIORITY_SCHEDULING Option Group.

signed-off-by: Gaetan Perrot <gaetanperrotpro@gmail.com>

moonlight83340 commented:
> @moonlight83340 - nice work 👍
> 
> I think there may be some additional tests to write, and likely some additional work on the implementation of `pthread_create()`
> 
> Tests:
> 
>     1. Verify that the scheduling attributes of the parent thread (policy, priority, ..) are inherited by a child thread when INHERIT_SCHED is used. This may be tricky, because currently Zephyr threads are not pthreads. It might be necessary to extrapolate equivalent pthread attributes (policy, priority) from the running `k_thread` and compare those.
> 
>     2. Verify that the scheduling attributes specified by the pthread_attr_t are passed to the child thread.
> 
> 
> For both of those, you should be able to create a static helper function in the same file. Something like this?
> 
> ```c++
> static void test_pthread_attr_set_inheritsched_common(int inheritsched);
> ```
@cfriedt I was trying something and just realize I can't modify the attr of the current thread ? 
Now I think I can understand the tricky part 😂 

I can modify the `pthread_attr_t` only before the thread is create, am I right ? 🤔

cfriedt commented on tests/posix/common/src/pthread_attr.c:
```diff
@@ -439,6 +439,114 @@ ZTEST(pthread_attr, test_pthread_attr_setstacksize)
+	zassert_equal(inheritsched, PTHREAD_INHERIT_SCHED);
+}
+
+static struct sched_param setinheritsched_sched_param;
+static int setinheritsched_policy;
+static int setinheritsched_inheritsched;
```
@moonlight83340 - would it make sense to simply pass the appropriate `pthread_attr_t *` via `arg` in `test_pthread_attr_set_inheritsched_child_fn()`?

cfriedt commented on tests/posix/common/src/pthread_attr.c:
```diff
@@ -439,6 +439,114 @@ ZTEST(pthread_attr, test_pthread_attr_setstacksize)
+	pthread_t parent;
+
+	zassert_ok(pthread_create(&parent, &attr,
+					  test_pthread_attr_set_inheritsched_parent_fn, NULL));
```
Does this work, @moonlight83340? With that, it should be possible to compare both the expected and actual values in `test_pthread_attr_set_inheritsched_parent_fn()`
```suggestion
					  test_pthread_attr_set_inheritsched_parent_fn, &attr));
```

moonlight83340 commented:
Not sure on the error, I will check that when I have time.

cfriedt commented:
@moonlight83340 - can you please update this PR?

moonlight83340 commented:
> @moonlight83340 - can you please update this PR?

I will take a look on the week ! ^^'

moonlight83340 commented:
> Assertion failed at CMAKE_SOURCE_DIR/src/pthread_attr.c:541: test_pthread_attr_set_inheritsched_parent_fn: pthread_create(&child, NULL, test_pthread_attr_set_inheritsched_child_fn, NULL) is non-zero

I think, I need to check on before the thread creation to found why the thread can't create another thread.

cfriedt commented on lib/posix/options/pthread.c:
```diff
@@ -599,6 +638,20 @@ int pthread_create(pthread_t *th, const pthread_attr_t *_attr, void *(*threadrou
+		if (self == NULL) {
+			K_SPINLOCK_BREAK;
+		}
+		if (self->attr.inheritsched == PTHREAD_INHERIT_SCHED) {
+			t->attr.priority = self->attr.priority;
+			t->attr.schedpolicy = self->attr.schedpolicy;
+			t->attr.inheritsched = self->attr.inheritsched;
+		}
```
Something like this could probably help, but it would be better to move it into the block ending on line 639.

```suggestion
		if (t->attr.inheritsched == PTHREAD_INHERIT_SCHED) {
     struct k_thread *self = k_current_get();

			/* There might be a function named posix_to_zephyr_priority() ? */
			t->attr.priority /* = ...(self) */;
			t->attr.schedpolicy = /* = ...(self) */;
		}
```

cfriedt reviewed (commented):
It's close - a suggestion to maybe simplify things

cfriedt commented:
@moonlight83340 - can you apply this patch on top of your existing code? (please also fix up the code and test commits seperately)
https://pastebin.com/a7x4GXmg

cfriedt commented:
@moonlight83340 - fixup commits are somewhat special. Check [this](https://blog.sebastian-daschner.com/entries/git-commit-fixup-autosquash) out. Also, the `Signed-off-by` line is automatically added to commit messages with `git commit --signoff` or `git commit -s`.

See also [Contribution Workflow](https://docs.zephyrproject.org/2.7.5/contribute/index.html#contribution-workflow).

moonlight83340 commented:
> @moonlight83340 - fixup commits are somewhat special. Check [this](https://blog.sebastian-daschner.com/entries/git-commit-fixup-autosquash) out. Also, the `Signed-off-by` line is automatically added to commit messages with `git commit --signoff` or `git commit -s`.
> 
> See also [Contribution Workflow](https://docs.zephyrproject.org/2.7.5/contribute/index.html#contribution-workflow).

Should I rebase ?  `git rebase --autosquash --interactive base`

cfriedt commented:
@moonlight83340 - this would be a more accurate test, I think.

https://pastebin.com/1mm733Lw

cfriedt reviewed (approved):
LGTM - thanks 👍

moonlight83340 commented:
> LGTM - thanks 👍

Thanks to you, I will carefully check your patch. 
Hope to get the chance to continue my contribution !
